scripts/Muon/GUI/Common/threading_manager.py

```diff
@@ -0,0 +1,282 @@
+    equally across a specified number of threads.
+
+    Example : for a function f(x,y) where x and y are ints, the WorkerManager allows for evaluation of
+    f(x=[1,2,3,4,5],y=[1,2,3,4,5]) by splitting the lists across the given number of threads. So for two threads the
+    following evaluations might occur:
+        Thread 1 : f(1,1), f(3,3), f(5,5)
+        Thread 2 : f(2,2), f(4,4)
+
+    A list of callbacks can be passed into the constructor to respond to events (all threads finishing,
+    threads updating on progress, threads catching exceptions from the function evaluation)
+    """
+
+    finished = Signal()  # used for unit tests
+    cancelled = Signal()  # used for unit tests
+
+    def __init__(self, fn, num_threads,
+                 callback_on_threads_complete=lambda: 0,
+                 callback_on_progress_update=lambda dbl: 0,
+                 callback_on_thread_exception=lambda err_dict: 0,
+                 verbose=0,
+                 **kwarg_list):
+        """
+
+        :param fn: The function to be evaluated.
+        :param num_threads: Number of threads to use.
+        :param callback_on_threads_complete: Optional callable for when all threads are finished
+        :param callback_on_progress_update: Optional callback for progress updates
+        :param callback_on_thread_exception: Optional callback called on exception from thread
+        :param verbose: 0,1 whether to print output
+        :param kwarg_list: Keyword argument lists to apply to function (e.g. x =[1,2,3] for f(x) where x is an integer)
+        """
+        super(WorkerManager, self).__init__()
+
+        # Callback called when all threads have executed
+        self.thread_complete_callback = callback_on_threads_complete
+        # Callback called when a single thread updates its progress
+        self.progress_callback = callback_on_progress_update
+        # Callback called when an error is thrown from a thread
+        self.error_callback = callback_on_thread_exception
+
+        self.fn = fn
+        self.kwarg_list = kwarg_list
+
+        # maintain a list of running threads to allow for cancelling
+        self._threads = []
+
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
+        self._failed_results = {key: [] for key in self.kwarg_list.keys()}
+
+        self._progress = 0.0
+
+        self._num_threads = num_threads  # total number of threads started
+        self._thread_count = 0  # current threads running
+
+        self.mutex = QtCore.QMutex()
+
+        self.verbose = verbose
+
+    def _cancel_threads(self):
+        for thread in self._threads:
+            if thread.isRunning():
+                thread.cancel()
+
+    def _clear_threads(self):
+        self._cancel_threads()
+        self._threads = []
+        self._thread_count = 0
+
+    def _clear_results(self):
+        self._progress = 0.0
+        self._results = {key: [] for key in self.kwarg_list.keys() + ['results']}
```

Comment:
Can you help me understand this? Is it storing the inputs and then an extra field for the results? 

Is there a need to store the results at all? Can we just provide them in an the optional callback given by the caller?

Reply:
Yes, the results are stored alongside all the inputs used to generate them as the threads will be asynchronously writing to `_results` so I need to be able to identify which result came from which evaluation.

I feel there is a need to store the results so that they can all be accessed at once after all the threads are finished; if you wanted to do that via a callback you would still need to keep the results somewhere no?